### ISSUE TITLE
Two `textDocument/publishDiagnostics` notifications being received where I would expect one.

### ISSUE THREAD

**Issue by martskins** [haskell/haskell-language-server#1636]
### Your environment
Output of `haskell-language-server --probe-tools` or `haskell-language-server-wrapper --probe-tools`:
```sh
haskell-language-server version: 1.0.0.0 (GHC: 8.10.4) (PATH: /home/martin/.ghcup/bin/haskell-language-server-wrapper-1.0.0) (GIT hash: 4cd1cf934638881e52b3eba9f70157a4b799c0e9)
Tool versions found on the $PATH
cabal:          3.4.0.0
stack:          Not found
ghc:            8.10.4
```
Which lsp-client do you use:
Vim 8.2 with LanguageClient-neovim as the actual LSP client.

Describe your project (alternative: link to the project): https://github.com/sim590/language-client-highlight-bug

### Context

I've taken the liberty of adding a context section here for this one as I believe it needs some of it. This is an issue that has been reported by a `LanguageClient-neovim` user (the original issue: https://github.com/autozimu/LanguageClient-neovim/issues/1210). The issue, essentially, is that some diagnostics are not being cleared, and I think that this is some race condition on the client, but while I was trying to understand what was happening I found that in this particular scenario we are receiving two publishDiagnostics notifications from HLS where I would expect a single one. The first notification usually has a typecheck error, and the second one has the error on the first one plus a couple of warnings from hlint.  I've detailed the steps to reproduce but taking a peek at the original issue is probably also worth it.

I guess what I'm trying to find out here is if that's intentional from HLS or if that's not expected. I suspect we might have to deal with on the client side somehow anyways, but I was surprised by these two notifications being received virtually at the same time and figured I'd ask what the reason for this might be here.

### Steps to reproduce
 - Clone the repo
 - cd language-client-highlight-bug
 - vim -u minimal-vimrc app/Main.hs
 - press `7Gf=wC undefined<ESC>`

### Include debug information
This is an example of the pair of notifications I get for this scenarrio, as I said above, the first one has the typecheck error, and the second one has the typecheck error + some warnings from hlint.

```
{
  "diagnostics":[
    {"code":"-Wdeferred-out-of-scope-variables","message":"Variable not in scope: un :: Int","range":{"end":{"character":10,"line":6},"start":{"character":8,"line":6}},"severity":1,"source":"typecheck"},
  ],
  "uri":"file:///home/martin/dev/language-client-highlight-bug/app/Main.hs",
  "version":3,
}

{
  "diagnostics":[
    {"code":"Use camelCase","message":"Use camelCase\\nFound:\\n  my_func :: Int -> Int\\nWhy not:\\n  myFunc :: Int -> Int\\n","range":{"end":{"character":21,"line":2},"start":{"character":0,"line":2}},"severity":3,"source":"hlint"},
    {"code":"Use camelCase","message":"Use camelCase\\nFound:\\n  my_func = ...\\nWhy not:\\n  myFunc = ...\\n","range":{"end":{"character":19,"line":3},"start":{"character":0,"line":3}},"severity":3,"source":"hlint"},
    {"code":"-Wdeferred-out-of-scope-variables","message":"Variable not in scope: un :: Int","range":{"end":{"character":10,"line":6},"start":{"character":8,"line":6}},"severity":1,"source":"typecheck"},
  ],
  "uri":"file:///home/martin/dev/language-client-highlight-bug/app/Main.hs",
  "version":3,
}
```

**Comment by pepeiborra** [comment]
HLS aggregates diagnostics over windows of 0.1s, and there are multiple warnings producers that run in parallel, so it seems perfectly plausible that two diagnostic notifications are emitted for the same file. 

If that creates problems, we can look at increasing the window to 0.2s or even more. 

Alternatively, the LSP specification has a provision for partial responses, although it was not implemented by any clients out there last time I checked

**Comment by martskins** [comment]
Actually, it's not really an issue anymore, I managed to place that logic behind a lock now so that won't be racing anymore, probably should've been like that in the first place anyways. I'm closing this as your response has answered my question and the original issue is solved now as well. Thank you!
